mro/custom_types.py

```diff
@@ -0,0 +1,214 @@
+                                LEFT JOIN pg_catalog.pg_namespace n
+                                       ON n.oid = t.typnamespace
+                                WHERE 
+                                (
+                                      t.typrelid = 0 
+                                      OR 
+                                      (
+                                           SELECT c.relkind = 'c' 
+                                           FROM pg_catalog.pg_class c
+                                           WHERE c.oid = t.typrelid
+                                      )
+                                ) 
+                                AND NOT EXISTS
+                                (
+                                      SELECT 1
+                                      FROM pg_catalog.pg_type el 
+                                      WHERE el.oid = t.typelem
+                                      AND el.typarray = t.oid
+                                )
+                                AND n.nspname NOT IN 
+                                (
+                                      'pg_catalog',
+                                      'information_schema'
+                                )
+                        """)
+    connection.commit()
+    types = [t for t in cursor]
+    return types
+
+
+def customColumnToDataType(column, code_start, code_end):
+    python_type = f'mro.custom_types.{column[27]}'
+    return '{0}{1}, {2}'.format(code_start, python_type, code_end)
+
+
+def strtobool(value: str):
+    return {'true': True,
+            'false': False,
+            't': True,
+            'f': False
+            }.get(value.lower(), None)
+
+
+def strtodate(value: str):
+    return datetime.datetime.strptime(value, '%Y-%m-%d').date()
+
+
+def strtotime(value: str):
+    return datetime.datetime.strptime(value, '%H:%M:%S.%f').time()
+
+
+def strtotimestamp(value: str):
+    return datetime.datetime.strptime(value, '%Y-%m-%d %H:%M:%S.%f')
+
+
+postgres_type_to_python_map = {
+    'character varying': str,
+    'text': str,
+    'integer': int,
+    'boolean': strtobool,
+    'bool': strtobool,
+    'float': float,
+    'double precision': float,
+    'real': float,
+    'date': strtodate,
+    'timestamp without time zone': strtotimestamp,
+    'time without time zone': strtotime,
+    'json': str,
+    'jsonb': str,
+    'uuid': str,
+    'bytea': str,
+}
+
+
+class CustomFieldType:
+    def __init__(self, attribute):
+        self.attribute = attribute
+
+    def __get__(self, instance, type):
+        if instance is None:
+            return self
+        else:
+            return attrgetter(self.attribute)(instance)
+
+    def __set__(self, instance, value):
+        raise NotImplementedError("You cannot set custom type internal attributes")
+
+
+def create_custom_types(connection):
+    custom_types = _get_custom_type_names(connection)
+
+    for custom_type in custom_types:
```

Comment:
This method essentially gets all the custom type names, registers them with psycopg2, creates a python class to represent them and a db_python class to represent them, as part of this it registers the classes with psycopg2 so it knows how to automatically convert between the string returned from postgres into the python class. When inserting into the db it needs to take the python class and use the db_python class to know how to run the sql.